##  1 Data wrangling<a id='1_Data_wrangling'></a>

### 1.1 Contents<a id='1.1_Contents'></a>
* [1 Data wrangling](#2_Data_wrangling)
  * [1.1 Contents](#2.1_Contents)
  * [1.2 Introduction to American Express Credit Default Project](#1.2_Introduction_to_American_Express_Credit_Default_Project)
    * [1.2.1 Brief Problem Description](#1.2.1_Brief_Problem_Description)
    * [1.2.2 Introduction To Notebook](#1.2.2_Introduction_To_Notebook)
  * [1.3 Imports](#1.3_Imports)
  * [1.4 Objectives](#1.4_Objectives)
  * [1.5 Data Collection](#2.5_Data_Collection)
  * [1.6 Data Exploration](#2.6_Data_Exploration)
  

#### 1.2 Introduction to American Express Credit Default Project<a id='1.2_Introduction to American Express Credit Default Project'></a>



##### 1.2.1 Brief Problem Description<a id='1.2.1_Brief Problem Description'></a>

##### 1.2.2 Introduction To Notebook<a id='1.2.2_Introduction To Notebook'></a>

##### 1.3 Imports<a id='1.3_Imports'></a>

In [20]:
# import of relevant packages
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import os

##### 1.5 Data Collection<a id='1.5_Data Collection'></a>

In [21]:
# Loading the csv file
#fn = os.path.join(os.path.dirname(__file__), 'train.csv')
amex_df = pd.read_csv('/Users/camilods16/Documents/Project-2-AmEx-Credit-Card-Default-/Project-2-AmEx-Credit-Card-Default-/data/raw/train.csv', sep=',')

##### 1.7 Data Definition<a id='1.7_Data Exploration'></a>

In [24]:
#calling info() on raw data to have a quick overview of the dataset components
amex_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45528 entries, 0 to 45527
Data columns (total 19 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customer_id              45528 non-null  object 
 1   name                     45528 non-null  object 
 2   age                      45528 non-null  int64  
 3   gender                   45528 non-null  object 
 4   owns_car                 44981 non-null  object 
 5   owns_house               45528 non-null  object 
 6   no_of_children           44754 non-null  float64
 7   net_yearly_income        45528 non-null  float64
 8   no_of_days_employed      45065 non-null  float64
 9   occupation_type          45528 non-null  object 
 10  total_family_members     45445 non-null  float64
 11  migrant_worker           45441 non-null  float64
 12  yearly_debt_payments     45433 non-null  float64
 13  credit_limit             45528 non-null  float64
 14  credit_limit_used(%)  

There are 19 columns with a range of 45528 entries. Some columns have different number of entries. The following are the column with less entries: 
- owns_car
- no_of_childer
- no_of_days_emplyoyed
- total_family_members
- migrant_worker
- yearly_debt_payments **
- credit_score **

some features stand out just by name: credit score, debt payments and all default-related. 
3 types in the dataset: float, int, object. 

There are columns does not seem to correspond the content stored in them such as no_of_children and migrant_worker. However, in the next steps I will apply methods to define and understan better the dataset.

In [23]:
amex_df.head()

customer_id            name  age gender owns_car owns_house  no_of_children  \
0  CST_115179        ita Bose   46      F        N          Y             0.0   
1  CST_121920  Alper Jonathan   29      M        N          Y             0.0   
2  CST_109330     Umesh Desai   37      M        N          Y             0.0   
3  CST_128288             Rie   39      F        N          Y             0.0   
4  CST_151355          McCool   46      M        Y          Y             0.0   

   net_yearly_income  no_of_days_employed occupation_type  \
0          107934.04                612.0         Unknown   
1          109862.62               2771.0        Laborers   
2          230153.17                204.0        Laborers   
3          122325.82              11941.0      Core staff   
4          387286.00               1459.0      Core staff   

   total_family_members  migrant_worker  yearly_debt_payments  credit_limit  \
0                   1.0             1.0              33070.28      18690.93   
1                   2.0             0.0              15329.53      37745.19   
2                   2.0             0.0              48416.60      41598.36   
3                   2.0             0.0              22574.36      32627.76   
4                   1.0             0.0              38282.95      52950.64   

   credit_limit_used(%)  credit_score  prev_defaults  default_in_last_6months  \
0                    73         544.0              2                        1   
1                    52         857.0              0                        0   
2                    43         650.0              0                        0   
3                    20         754.0              0                        0   
4                    75         927.0              0                        0   

   credit_card_default  
0                    1  
1                    0  
2                    0  
3                    0  
4                    0

In [28]:
# summary statistics for each column
amex_df.describe().T

count           mean            std       min  \
age                      45528.0      38.993411       9.543990     23.00   
no_of_children           44754.0       0.420655       0.724097      0.00   
net_yearly_income        45528.0  200655.622249  669074.034607  27170.61   
no_of_days_employed      45065.0   67609.289293  139323.524434      2.00   
total_family_members     45445.0       2.158081       0.911572      1.00   
migrant_worker           45441.0       0.179111       0.383450      0.00   
yearly_debt_payments     45433.0   31796.965311   17269.727234   2237.47   
credit_limit             45528.0   43548.416028  148784.686878   4003.14   
credit_limit_used(%)     45528.0      52.235020      29.376910      0.00   
credit_score             45520.0     782.791257     100.619746    500.00   
prev_defaults            45528.0       0.060710       0.264629      0.00   
default_in_last_6months  45528.0       0.050540       0.219059      0.00   
credit_card_default      45528.0       0.081203       0.273149      0.00   

                                25%         50%          75%           max  
age                          31.000      39.000      47.0000  5.500000e+01  
no_of_children                0.000       0.000       1.0000  9.000000e+00  
net_yearly_income        126345.835  171714.910  240603.7600  1.407590e+08  
no_of_days_employed         936.000    2224.000    5817.0000  3.652520e+05  
total_family_members          2.000       2.000       3.0000  1.000000e+01  
migrant_worker                0.000       0.000       0.0000  1.000000e+00  
yearly_debt_payments      19231.140   29081.650   40561.1500  3.281129e+05  
credit_limit              23973.805   35688.045   53435.7625  3.112997e+07  
credit_limit_used(%)         27.000      54.000      78.0000  9.900000e+01  
credit_score                704.000     786.000     867.0000  9.490000e+02  
prev_defaults                 0.000       0.000       0.0000  2.000000e+00  
default_in_last_6months       0.000       0.000       0.0000  1.000000e+00  
credit_card_default           0.000       0.000       0.0000  1.000000e+00

In [34]:
amex_df[['net_yearly_income']].median()

net_yearly_income    171714.91
dtype: float64

In [37]:
amex_df.median()

/var/folders/fk/kvpxv24d2ys_slgf4kqk23f40000gn/T/ipykernel_3458/413277450.py:1: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  amex_df.median()


age                            39.000
no_of_children                  0.000
net_yearly_income          171714.910
no_of_days_employed          2224.000
total_family_members            2.000
migrant_worker                  0.000
yearly_debt_payments        29081.650
credit_limit                35688.045
credit_limit_used(%)           54.000
credit_score                  786.000
prev_defaults                   0.000
default_in_last_6months         0.000
credit_card_default             0.000
dtype: float64

In [32]:
# missing values per column
missing_val = pd.concat([amex_df.isnull().sum(), 100 * amex_df.isnull().mean()], axis=1)
missing_val.columns=['count', '%']
missing_val.sort_values(by='count', ascending=False)

count         %
no_of_children             774  1.700053
owns_car                   547  1.201458
no_of_days_employed        463  1.016957
yearly_debt_payments        95  0.208663
migrant_worker              87  0.191091
total_family_members        83  0.182305
credit_score                 8  0.017572
customer_id                  0  0.000000
default_in_last_6months      0  0.000000
prev_defaults                0  0.000000
credit_limit_used(%)         0  0.000000
credit_limit                 0  0.000000
occupation_type              0  0.000000
name                         0  0.000000
net_yearly_income            0  0.000000
owns_house                   0  0.000000
gender                       0  0.000000
age                          0  0.000000
credit_card_default          0  0.000000

Small percentage of missingness. The top 3 are missing 1% to 1.7% of their data. no_of_children is missing data and its type (float) does not correspond to its definition(int):

no_of_children, owns_car, no_of_days_employed, yearly_debt_payments, migrant_worker, total_family_members.

- Is there 'hiden' missing data? Are all variables relevant to solve the problem? What do we need from the dataset to solve the business problem? What are potential features for analysis? To explore questions further I will separate the categorical and numerical variables.

Categorical Variables:

In [45]:
categorical_v = amex_df.select_dtypes(include='object')
categorical_v

customer_id            name gender owns_car owns_house occupation_type
0      CST_115179        ita Bose      F        N          Y         Unknown
1      CST_121920  Alper Jonathan      M        N          Y        Laborers
2      CST_109330     Umesh Desai      M        N          Y        Laborers
3      CST_128288             Rie      F        N          Y      Core staff
4      CST_151355          McCool      M        Y          Y      Core staff
...           ...             ...    ...      ...        ...             ...
45523  CST_130421           Doris      F        N          N         Unknown
45524  CST_136670         Luciana      F        N          Y     Accountants
45525  CST_145435         Jessica      F        N          Y      Core staff
45526  CST_130913           Tessa      M        Y          N        Laborers
45527  CST_160078        Gopinath      M        N          Y         Unknown

[45528 rows x 6 columns]